Description: Fix segfault with zlib acceleration
 With zlib acceleration enabled gzip would segfault when
 given multiple files larger than 5KB.
Origin: upstream, https://git.savannah.gnu.org/cgit/gzip.git/commit/?id=be0a534ba2b6e77da289de8da79e70843b1028cc
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/zlib/+bug/1901528
Last-Update: 2021-01-15
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: gzip-1.10/dfltcc.c
===================================================================
--- gzip-1.10.orig/dfltcc.c
+++ gzip-1.10/dfltcc.c
@@ -235,10 +235,8 @@ dfltcc_gdht (struct dfltcc_param_v0 *par
   dfltcc (DFLTCC_GDHT, param, NULL, NULL, &next_in, &avail_in, NULL);
 }
 
-static off_t total_in;
-
 static dfltcc_cc
-dfltcc_cmpr_xpnd (struct dfltcc_param_v0 *param, int fn)
+dfltcc_cmpr_xpnd (struct dfltcc_param_v0 *param, int fn, off_t *total_in)
 {
   uch *next_out = outbuf + outcnt;
   size_t avail_out = OUTBUFSIZ - outcnt;
@@ -250,7 +248,7 @@ dfltcc_cmpr_xpnd (struct dfltcc_param_v0
                          window);
   off_t consumed_in = next_in - (inbuf + inptr);
   inptr += consumed_in;
-  total_in += consumed_in;
+  *total_in += consumed_in;
   outcnt += ((OUTBUFSIZ - outcnt) - avail_out);
   return cc;
 }
@@ -342,6 +340,7 @@ dfltcc_deflate (int pack_level)
 
   union aligned_dfltcc_param_v0 ctx_v0;
   struct dfltcc_param_v0 *param = init_param (&ctx_v0);
+  off_t total_in = 0;
 
   /* Compress ifd into ofd in a loop.  */
   while (true)
@@ -391,7 +390,8 @@ dfltcc_deflate (int pack_level)
         }
 
       /* Compress inbuf into outbuf.  */
-      while (dfltcc_cmpr_xpnd (param, DFLTCC_CMPR) == DFLTCC_CC_AGAIN)
+      while (dfltcc_cmpr_xpnd (param, DFLTCC_CMPR, &total_in)
+             == DFLTCC_CC_AGAIN)
         ;
 
       /* Unmask the input data.  */
@@ -420,6 +420,7 @@ dfltcc_inflate (void)
 
   union aligned_dfltcc_param_v0 ctx_v0;
   struct dfltcc_param_v0 *param = init_param (&ctx_v0);
+  off_t total_in = 0;
 
   /* Decompress ifd into ofd in a loop.  */
   while (true)
@@ -439,7 +440,8 @@ dfltcc_inflate (void)
 
         /* Decompress inbuf into outbuf.  */
         dfltcc_cc cc;
-        while ((cc = dfltcc_cmpr_xpnd (param, DFLTCC_XPND)) == DFLTCC_CC_AGAIN)
+        while ((cc = dfltcc_cmpr_xpnd (param, DFLTCC_XPND, &total_in))
+               == DFLTCC_CC_AGAIN)
           ;
         if (cc == DFLTCC_CC_OK)
           {
